jsettlers.logic/src/main/java/jsettlers/logic/map/grid/MainGrid.java

```diff
@@ -1801,6 +1854,109 @@ public final boolean isInBounds(ShortPoint2D position) {
 			return MainGrid.this.isInBounds(position.x, position.y);
 		}
 
+		public FerryEntrance ferryAtPosition(ShortPoint2D position, byte playerId) {
+			if (!isWaterSafe(position.x, position.y)) {
+				return null;
+			}
+			int x = position.x;
+			int y = position.y;
+			byte dx[] = EDirection.getXDeltaArray();
+			byte dy[] = EDirection.getYDeltaArray();
+			for (int i = -1; i < EDirection.NUMBER_OF_DIRECTIONS; i++) { // search ferry
+				if (i >= 0) {
+					x = position.x + dx[i];
+					y = position.y + dy[i];
+				}
+				ILogicMovable ship = getMovableGrid().getMovableAt(x, y);
+				if (ship.getMovableType() == EMovableType.FERRY && ship.getPlayer().playerId == playerId) {
+					for (int j = 0; j < EDirection.NUMBER_OF_DIRECTIONS; j++) { // search ferry entrance
+						if (!this.isBlocked(x + 2 * dx[j], y + 2 * dy[j])) {
+							FerryEntrance boarding = new FerryEntrance((Movable) ship,
+									new ShortPoint2D(x + 2 * dx[j], y + 2 * dy[j]));
+							return boarding;
+						}
+					}
+				}
+			}
+			return null;
+		}
+
+		public ShortPoint2D getUnloadPosition(ShortPoint2D position) {
+			int x = position.x;
+			int y = position.y;
+			int dx = 0;
+			int dy = 0;
+			final byte[] xDeltaArray = EDirection.getXDeltaArray();
+			final byte[] yDeltaArray = EDirection.getYDeltaArray();
+			int distance = 2;
+			int direction;
+			boolean searching = true;
+			for (direction = 0; direction < EDirection.NUMBER_OF_DIRECTIONS; direction++) {
+				dx = (short) (xDeltaArray[direction] * distance) + x;
+				dy = (short) (yDeltaArray[direction] * distance) + y;
+				if (isInBounds(new ShortPoint2D(dx, dy)) && !isWaterSafe(dx, dy)
+						&& !landscapeGrid.isBlocked(dx, dy)) {
+					searching = false;
+					break;
+				}
+			}
+			if (searching) {
+				return null; // no coast near by
+			} else {
+				return new ShortPoint2D(dx, dy); // coast found, return point for unloading
```

Comment:
A return inside the loop is as bad as a break. Don't complicate your code too much ;-).